File: packages/septr-python/septr/core/headers.py
"""Advisory security-header checks for runtime responses.

Reports responses missing standard security headers (CSP, HSTS, nosniff,
clickjacking, referrer). Detection-only — Septr never injects headers, since
values like CSP are app-specific and wrong guesses break apps.
"""

import time
from .secrets import DetectionEvent

SECURITY_HEADER_CHECKS: list[tuple[str, str, str]] = [
    ("Content-Security-Policy", "Content-Security-Policy header missing", "medium"),
    ("Strict-Transport-Security", "Strict-Transport-Security (HSTS) header missing", "medium"),
    ("X-Content-Type-Options", "X-Content-Type-Options header missing", "medium"),
    ("X-Frame-Options", "X-Frame-Options header missing", "medium"),
    ("Referrer-Policy", "Referrer-Policy header missing", "low"),
]


def detect_missing_security_headers(headers: list[tuple[bytes, bytes]]) -> list[DetectionEvent]:
    """Given raw response headers (bytes pairs), report the missing ones."""
    present = {
        key.decode("utf-8", "replace").lower()
        for key, _ in headers
    }
    events: list[DetectionEvent] = []
    for name, description, severity in SECURITY_HEADER_CHECKS:
        if name.lower() not in present:
            events.append(DetectionEvent(
                type="security_headers", severity=severity,
                patternId="missing_security_header",
                description=description,
                statusCode=200, timestamp=time.time() * 1000,
            ))
    return events

File: packages/septr-go/headers.go
package septr

import (
	"net/http"
)

var securityHeaderChecks = []struct {
	name        string
	description string
	severity    string
}{
	{"Content-Security-Policy", "Content-Security-Policy header missing", "medium"},
	{"Strict-Transport-Security", "Strict-Transport-Security (HSTS) header missing", "medium"},
	{"X-Content-Type-Options", "X-Content-Type-Options header missing", "medium"},
	{"X-Frame-Options", "X-Frame-Options header missing", "medium"},
	{"Referrer-Policy", "Referrer-Policy header missing", "low"},
}

// DetectMissingSecurityHeaders reports responses missing standard security
// headers. Advisory only — Septr never injects headers.
func DetectMissingSecurityHeaders(h http.Header) []DetectionEvent {
	present := map[string]bool{}
	for name := range h {
		present[lowerName(name)] = true
	}
	var events []DetectionEvent
	for _, c := range securityHeaderChecks {
		if !present[lowerName(c.name)] {
			events = append(events, DetectionEvent{
				Type: "security_headers", Severity: c.severity,
				PatternID: "missing_security_header", Description: c.description,
				StatusCode: 200, Timestamp: nowMs(),
			})
		}
	}
	return events
}

func lowerName(s string) string {
	b := make([]byte, len(s))
	for i := 0; i < len(s); i++ {
		c := s[i]
		if c >= 'A' && c <= 'Z' {
			c += 32
		}
		b[i] = c
	}
	return string(b)
}
